src/ParallelAlgorithms/LinearSolver/Gmres/ElementActions.hpp

```diff
@@ -313,18 +416,26 @@ struct NormalizeOperandAndUpdateField {
             const gsl::not_null<db::item_type<fields_tag>*> field,
             const gsl::not_null<size_t*> iteration_id,
             const gsl::not_null<Convergence::HasConverged*> local_has_converged,
-            const db::const_item_type<initial_fields_tag>&
-                initial_field) noexcept {
-          *operand /= normalization;
+            const db::const_item_type<initial_fields_tag>& initial_field,
```

Comment:
I think we're trying to stop using item_type everywhere. Can you replace with actual types or `auto&`?

Reply:
Are you suggesting to transition from item_type to auto because the latter may compile faster, or because item_type relies on our implementation which may not be robust enough to handle all possible cases, or is there a different reason?

Reply:
I'm suggesting to switch to using concrete types whenever at all possible, even if you have to do some TMP and define a type alias. E.g. mortar things could do something like `using MortarData = dg::MortarData<typename boundary_scheme::mortar_tags>;` and then use the type alias. Short of that, `auto` is no more or less helpful than `const db::const_item_type<elliptic::LinearSolver::Tags::Prefix<domain::Tags::Something<Metavariables::volume_dim>>>&`. Honestly, I would find `const auto&` easier to read just because I don't get a large vomit of "information" that requires me to hunt around the entire code base to figure out the type.

Reply:
Do you think we should also change the mutating arguments, i.e. `const gsl::not_null<db::item_type<Tag>*> arg` to `const auto arg`, even though the latter does not constrain the type to `gsl::not_null`?

Reply:
@nilsdeppe ping on this question

Reply:
@nilsdeppe ping on this question

Reply:
Yes, you should either use the type explicitly if you know it, or use `auto`. Personally, if I am using `auto` I often use the `_ptr` suffix. While not ideal, it reduces the ambiguity. Encoding type info into the name is not ideal, but until C++20 we can't do something like `gsl::not_null<T>` in a lambda. I find that `db::item_type` is no more useful than `auto`, and will make it harder for people to transition out of SpECTRE code. The fewer magic internal things we use everywhere the better, especially if there's something in the language that allows us to express exactly what we mean.

Reply:
Another tell that something's a `not_null` is it's `auto` and not `auto&`

Reply:
Ok sounds good, I'm on board with that 👍 Can't wait for the day we can refactor our code to C++20 😄

Reply:
:+1: agreed, I wish the explicit template parameters for lambdas was in 17 because that would make this code a lot easier to read :(

Reply:
[Abbreviated function templates](https://en.cppreference.com/w/cpp/language/function_template#Abbreviated_function_template) will also be very nice :)